consensus/istanbul/backend/announce.go

```diff
@@ -589,130 +649,311 @@ func (sb *Backend) regossipAnnounce(msg *istanbul.Message, payload []byte) error
+}
 
-		sb.cachedAnnounceMsgsMu.RLock()
-		if cachedEntry, ok := sb.cachedAnnounceMsgs[announceVersion.ValAddress]; !ok || cachedEntry.MsgVersion < announceVersion.AnnounceMsgVersion {
-			announcesToRequest = append(announcesToRequest, announceVersion.ValAddress)
+func (sb *Backend) upsertSignedAnnounceVersionEntries(entries []*vet.SignedAnnounceVersionEntry) error {
+	logger := sb.logger.New("func", "upsertSignedAnnounceVersions")
+	newEntries, err := sb.signedAnnounceVersionTable.Upsert(entries)
+	if err != nil {
+		logger.Warn("Error in upserting entries", "err", err)
+	}
+
+	// If this node is a validator (checked later as a result of this call) and it receives a signed announce
+	// version from a remote validator that is newer than the remote validator's
+	// version in the val enode table, this node did not receive a direct announce
+	// and needs to announce its own enode to the remote validator.
+	sb.startGossipAnnounceTask()
+
+	// Only regossip entries that are new to this node and do not originate
+	// from an address that we have gossiped a signed announce version for
+	// within the last 5 minutes
+	var signedAnnVersionsToRegossip []*signedAnnounceVersion
+	sb.lastSignedAnnounceVersionsGossipedMu.Lock()
+	for _, entry := range newEntries {
+		lastGossipTime, ok := sb.lastSignedAnnounceVersionsGossiped[entry.Address]
+		if !ok || time.Since(lastGossipTime) >= 5*time.Minute {
+			signedAnnVersionsToRegossip = append(signedAnnVersionsToRegossip, &signedAnnounceVersion{
+				Address: entry.Address,
+				Version: entry.Version,
+				Signature: entry.Signature,
+			})
+			sb.lastSignedAnnounceVersionsGossiped[entry.Address] = time.Now()
 		}
-		sb.cachedAnnounceMsgsMu.RUnlock()
 	}
+	sb.lastSignedAnnounceVersionsGossipedMu.Unlock()
+	if len(signedAnnVersionsToRegossip) > 0 {
+		return sb.gossipSignedAnnounceVersionsMsg(signedAnnVersionsToRegossip)
+	}
+	return nil
+}
+
+func (sb *Backend) updateAnnounceVersion() {
+	sb.updateAnnounceVersionCh <- struct{}{}
+	<-sb.updateAnnounceVersionCompleteCh
+}
 
-	if len(announcesToRequest) > 0 {
-		announcesToRequestBytes, err := rlp.EncodeToBytes(announcesToRequest)
+// setAndShareUpdatedAnnounceVersion generates a new enode certificate message and sends
+// it to this node's proxy (if applicable).
+// TODO: When a generated announce no longer creates a new version, the version
+// parameter can be removed and instead have the version generated in this function.
+// Tracked here: https://github.com/celo-org/celo-monorepo/issues/2668
+func (sb *Backend) setAndShareUpdatedAnnounceVersion(version uint) error {
+	logger := sb.logger.New("func", "updateAnnounceVersion")
+	// Send new versioned enode msg to all other registered or elected validators
+	validatorConnSet, err := sb.retrieveValidatorConnSet()
+	if err != nil {
+		return err
+	}
+	// Don't send any messages if this node is not registered or elected
+	if !validatorConnSet[sb.Address()] {
```

Comment:
This seems to be the same idea, but slightly different than the `shouldGenerateAndProcessAnnounce` function. Consider using that function or the same implementation here.